lib/script.ex

```diff
@@ -0,0 +1,761 @@
+  """
+  @spec serialize_script(t()) :: binary
+  def serialize_script(script = %__MODULE__{}) do
+    # serialize_script(%Script{items: [0x81]}) will still display "Q" but 
+    # it functions as binary 0x51. Use to_hex for displaying scripts.
+    serializer(script, <<>>)
+  end
+
+  @doc """
+  	to_hex returns the hex of a serialized script. 
+  """
+  @spec to_hex(t()) :: String.t()
+  def to_hex(script) do
+    script
+    |> serialize_script()
+    |> Base.encode16(case: :lower)
+  end
+
+  @doc """
+  	parse_script parses a binary or hex string into a script.
+  """
+  @spec parse_script(binary) :: {:ok, t()} | {:error, String.t()}
+  def parse_script(script_str) when is_binary(script_str) do
+    try do
+      case Utils.hex_to_bin(script_str) do
+        {:error, _msg} ->
+          # necessary to allow parse_script to accept raw binary script
+          parser(new(), script_str)
+
+        bin ->
+          parser(new(), bin)
+      end
+    rescue
+      _ -> {:error, "invalid script. parse_script accepts hex or binary."}
+    end
+  end
+
+  defp parser(script, <<>>), do: {:ok, script}
+
+  defp parser(script, <<next::binary-size(1), bin::binary>>) do
+    op = :binary.decode_unsigned(next)
+
+    cond do
+      # PUSHBYTES
+      op > 0x00 and op < 0x4C ->
+        {:ok, rest} = parser(script, :binary.part(bin, op, byte_size(bin) - op))
+
+        rest
+        |> push_raw_data(:binary.part(bin, 0, op))
+        |> push_op(op)
+
+      # PUSHDATA1
+      op == 0x4C ->
+        len = bin |> :binary.part(0, 1) |> Utils.little_to_int()
+        {:ok, rest} = parser(script, :binary.part(bin, len + 1, byte_size(bin) - len - 1))
+
+        rest
+        |> push_raw_data(:binary.part(bin, 1, len))
+        |> push_op(op)
+
+      # PUSHDATA2
+      op == 0x4D ->
+        len = bin |> :binary.part(0, 2) |> Utils.little_to_int()
+        {:ok, rest} = parser(script, :binary.part(bin, len + 2, byte_size(bin) - len - 2))
+
+        rest
+        |> push_raw_data(:binary.part(bin, 2, len))
+        |> push_op(op)
+
+      # PUSHDATA4
+      op == 0x4E ->
+        len = bin |> :binary.part(0, 4) |> Utils.little_to_int()
+        {:ok, rest} = parser(script, :binary.part(bin, len + 4, byte_size(bin) - len - 4))
+
+        rest
+        |> push_raw_data(:binary.part(bin, 4, len))
+        |> push_op(op)
+
+      # OPCODE
+      is_valid_opcode(op) ->
+        {:ok, rest} = parser(script, :binary.part(bin, 0, byte_size(bin)))
+        push_op(rest, op)
+
+      true ->
+        invalid_opcode_error(op)
+    end
+  end
+
+  @doc """
+  	raw_combine directly concatenates two scripts with no checks.
+  """
+  @spec raw_combine(t(), t()) :: t()
+  def raw_combine(%__MODULE__{items: s1}, %__MODULE__{items: s2}),
+    do: %__MODULE__{items: s1 ++ s2}
+
+  @doc """
+  	display_script returns a human readable string of the script, with
+  	op_codes shown by name rather than number. 
+  """
+  @spec display_script(t()) :: String.t()
```

Comment:
maybe we should implement String.Char protocol using this function. I don't think we need to do this in this PR though